ktbridge/plugin/src/main/kotlin/app/cash/quickjs/ktbridge/plugin/BridgedCallRewriter.kt

```diff
@@ -0,0 +1,327 @@
+import org.jetbrains.kotlin.ir.ObsoleteDescriptorBasedAPI
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.builders.IrBuilderWithScope
+import org.jetbrains.kotlin.ir.builders.declarations.addValueParameter
+import org.jetbrains.kotlin.ir.builders.declarations.buildFun
+import org.jetbrains.kotlin.ir.builders.irBranch
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.builders.irEquals
+import org.jetbrains.kotlin.ir.builders.irGet
+import org.jetbrains.kotlin.ir.builders.irString
+import org.jetbrains.kotlin.ir.builders.irTrue
+import org.jetbrains.kotlin.ir.builders.irWhen
+import org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin
+import org.jetbrains.kotlin.ir.declarations.IrFactory
+import org.jetbrains.kotlin.ir.declarations.IrField
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.IrBranch
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.IrExpressionBody
+import org.jetbrains.kotlin.ir.expressions.IrFunctionExpression
+import org.jetbrains.kotlin.ir.expressions.IrGetValue
+import org.jetbrains.kotlin.ir.expressions.IrReturn
+import org.jetbrains.kotlin.ir.expressions.IrStatementOrigin
+import org.jetbrains.kotlin.ir.expressions.impl.IrBlockBodyImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrFunctionExpressionImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrReturnImpl
+import org.jetbrains.kotlin.ir.symbols.IrSimpleFunctionSymbol
+import org.jetbrains.kotlin.ir.symbols.impl.IrSimpleFunctionSymbolImpl
+import org.jetbrains.kotlin.ir.types.classFqName
+import org.jetbrains.kotlin.ir.types.defaultType
+import org.jetbrains.kotlin.ir.types.typeWith
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.getPropertyGetter
+import org.jetbrains.kotlin.ir.util.irCall
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+/**
+ * Rewrites calls to `createBridgeToJs` with two arguments:
+ *
+ * ```
+ * val helloService = createBridgeToJs(JsEchoService("hello"), EchoJsAdapter)
+ * ```
+ *
+ * to the overload that takes three arguments:
+ *
+ * ```
+ * val helloService = createBridgeToJs(
+ *   service = JsEchoService("hello"),
+ *   jsAdapter = EchoJsAdapter,
+ *   block = fun (bridgedCall: BridgedCall<JsEchoService>): ByteArray {
+ *     return when {
+ *       bridgedCall.funName == "echo" -> {
+ *         bridgedCall.result(
+ *           bridgedCall.service.echo(
+ *             bridgedCall.parameter()
+ *           )
+ *         )
+ *       }
+ *       else -> bridgedCall.unexpectedFunction()
+ *     }
+ *   }
+ * )
+ * ```
+ *
+ * This synthesizes a function that routes calls by their name, decodes parameters, and encodes the
+ * result.
+ */
+@ObsoleteDescriptorBasedAPI // TODO(jwilson): is there an alternative?
```

Comment:
Yes, replace usages of `descriptor` with `owner` instead. Then you won't need to translate types either.

Reply:
Yeah that was easier than I expected!
https://github.com/cashapp/quickjs-java/pull/231